### ISSUE TITLE
vgp accuracy 

### ISSUE THREAD

**Issue by hughsalimbeni** [GPflow/GPflow#389]
While the current method equivalence test passes, I've noticed that vgp becomes quite inaccurate when N is large. I find that the svgp/fitc/sgpr methods with Z=X are pretty close to gpr with a Gaussian likelihood for any N in the hundreds, but not so with vgp. Could the tests be made more stringent?

e.g. here is an example modified from the testing code with fixed hypers and more points:

```python
Ns = np.arange(1, 20)*10
Ls = [[] for _ in range(6)]

for N in Ns:
    X = np.linspace(-10, 10, N)[:, None]
    Y = np.sin(X) + 0.9 * np.cos(X*1.6) + np.random.randn(*X.shape) * 0.8

    m1 = GPR(X, Y, RBF(1))
    m2 = VGP(X, Y, RBF(1), Gaussian())
    m3 = SVGP(X, Y, RBF(1), Gaussian(), X.copy(), q_diag=False, whiten=False)
    m4 = SVGP(X, Y, RBF(1), Gaussian(), X.copy(), q_diag=False, whiten=True)
    m5 = SGPR(X, Y, RBF(1), X.copy())
    m6 = GPRFITC(X, Y, RBF(1), X.copy())

    models = [m1, m2, m3, m4, m5, m6]
    
    for m, L in zip(models, Ls):
        m.kern.fixed = True
        m.likelihood.fixed = True
        if hasattr(m, 'Z'):
            m.Z.fixed = True
        if len(m.get_free_state())>1:
            m.optimize()
                
        L.append(float(m.compute_log_likelihood()))
        
Ls = [np.array(L) for L in Ls]

plt.plot(Ns, Ls[0]-Ls[1], label='vgp')
#plt.plot(Ns, Ls[0]-Ls[2], label='svgp') # very wrong for N>30
plt.plot(Ns, Ls[0]-Ls[3], label='svgp white')
plt.plot(Ns, Ls[0]-Ls[4], label='sgpr')
plt.plot(Ns, Ls[0]-Ls[5], label='fitc')

plt.title('Difference in log marginal likelihood from exact model')
plt.axes().set_xlabel('number points')
plt.axes().set_ylabel('error')
plt.savefig('err.pdf', bbox_inches = 'tight')
plt.legend(loc='upper left', bbox_to_anchor=(1., 1.))
plt.show()

```
[err.pdf](https://github.com/GPflow/GPflow/files/873219/err.pdf)
(svgp and sgpr are indistinguishable on this plot)

**Comment by hughsalimbeni** [comment]
Also, I've just played around with smaller values of N and it seems vgp is very accurate for N<20 and fairly accurate for N<40, which is why it's not showing up on the test case (which is N=20)

**Comment by jameshensman** [comment]
Hi @hughsalimbeni , 

Nice observation. The VGP class in GPflow implements the Opper and Archambeau 2009 paper, where the number of free variables is reduced to 2N, rather than O(N^2). This seems like a cool idea at first, but I've also noticed some strange things in practise. I think perhaps a more investigation is needed. 

It would be straight-forward enough to implement a VGP that didn't have the reparameterization -- which is equivalent to the methods you described, but a bespoke implementation would avoid a couple of unnecessary solves I think. Here's a gist:

```python
import GPflow
import tensorflow as tf
import numpy as np
from GPflow.param import DataHolder
from GPflow._settings import settings
float_type = settings.dtypes.float_type


class VGP(GPflow.model.GPModel):
    def __init__(self, X, Y, kern, likelihood,
                 mean_function=GPflow.mean_functions.Zero(), num_latent=None):
        X = DataHolder(X, on_shape_change='recompile')
        Y = DataHolder(Y, on_shape_change='recompile')
        GPflow.model.GPModel.__init__(self, X, Y, kern, likelihood, mean_function)
        self.num_data = X.shape[0]
        self.num_latent = num_latent or Y.shape[1]
        self.q_mean = GPflow.param.Param(np.zeros((self.num_data, self.num_latent)))
        q_sqrt = np.array([np.eye(self.num_data)
                           for _ in range(self.num_latent)]).swapaxes(0, 2)
        self.q_sqrt = GPflow.param.Param(q_sqrt)

    def build_likelihood(self):
        K = self.kern.K(self.X) + tf.eye(self.num_data, dtype=float_type) * 1e-6
        L = tf.cholesky(K)

        # get mean and variance of f
        tmp = [tf.matmul(L, self.q_sqrt[:, :, i]) for i in range(self.num_latent)]
        f_var = tf.concat([tf.reduce_sum(tf.square(tmp_i), axis=[1], keep_dims=True) for tmp_i in tmp], axis=1)
        f_mean = self.q_mean + self.mean_function(self.X)

        KL = GPflow.kullback_leiblers.gauss_kl(self.q_mean, self.q_sqrt, K)

        var_exp = self.likelihood.variational_expectations(f_mean, f_var, self.Y)

        return tf.reduce_sum(var_exp) - KL


    def build_predict(self, Xnew, full_cov=False):
        return conditionals.conditional(Xnew, self.X, self.kern, self.q_mean,
                                        q_sqrt=self.q_sqrt, full_cov=full_cov, whiten=False)
```

Perhaps you could plug that into your code above and see what happens?

PS. write `python` after three back-ticks to get syntax highlighted code on github.

**Comment by hughsalimbeni** [comment]
Thanks for your reply @jameshensman 

I think there may be an extraneous L in the code above as the representation is unwhitened, and also the KL and conditional functions force lower triangular, so it should match. With these modifications it might be something like:

``` python

def build_likelihood(self):
    # mean of f
    f_mean = self.q_mean + self.mean_function(self.X)
    
    # var of f
    Lq = tf.matrix_band_part(tf.transpose(self.q_sqrt, (2, 0, 1)), -1, 0)  # force lower triangle DNN
    q_var_diag = [tf.reduce_sum(Lq[d]**2, 1) for d in range(self.num_latent)] #D[N]
    # or slower
    # q_var = [tf.matmul(Lq[d], Lq[d], transpose_b=True) for d in range(self.num_latent)] #D[NN]
    # q_var_diag = [tf.diag_part(q_var_i) for q_var_i in q_var] # D[N]
    f_var = tf.transpose(tf.stack(q_var_diag)) # ND

    # E_q(f) log p(y|f)        
    var_exp = self.likelihood.variational_expectations(f_mean, f_var, self.Y)
    
    # KL term         
    K = self.kern.K(self.X) + tf.eye(self.num_data, dtype=float_type) * 1e-6
    KL = GPflow.kullback_leiblers.gauss_kl(self.q_mean, self.q_sqrt, K)
    return tf.reduce_sum(var_exp) - KL
```
Comparing this to the other models it looks a lot like the unwhitened svgp. I also experimented with initialising q_sqrt to chol(K(X,X)) but this didn't seem to make a difference.

We could also do a whitened version (I called the previous one `VGP_full`)
``` python
 class VGP_white(VGP_full):
    def build_likelihood(self):
        K = self.kern.K(self.X) + tf.eye(self.num_data, dtype=float_type) * 1e-6
        L = tf.cholesky(K)

        # mean of f
        f_mean = tf.matmul(L, self.q_mean) + self.mean_function(self.X)
        
        # variance of f
        Lq = tf.matrix_band_part(tf.transpose(self.q_sqrt, (2, 0, 1)), -1, 0)  # DNN
        tmp = [tf.matmul(L, Lq[i]) for i in range(self.num_latent)] # D[NN]
        f_var = tf.concat([tf.reduce_sum(tf.square(tmp_i), axis=[1], keep_dims=True) for tmp_i in tmp], axis=1)

        # E log p(y|f)        
        var_exp = self.likelihood.variational_expectations(f_mean, f_var, self.Y)

        # KL term         
        KL = GPflow.kullback_leiblers.gauss_kl_white(self.q_mean, self.q_sqrt)
        
        return tf.reduce_sum(var_exp) - KL
        
    def build_predict(self, Xnew, full_cov=False):
        return conditional(Xnew, self.X, self.kern, self.q_mean,
                           q_sqrt=self.q_sqrt, full_cov=full_cov, whiten=True)
```

Running this through the code above this version is not identical to `svgp white`\`sgpr', but is pretty close (and appears closer to the ground truth in this example). The error certainly isn't on the same scale as `vgp'

My suspicion was that there might be a bug in the `vgp' implementation somewhere, but I've just compared to a slow implementation:

``` python
def inv(M):
    I = tf.eye(tf.shape(M)[0], dtype=tf.float64)
    L = tf.cholesky(M + I * 1e-6)
    return tf.cholesky_solve(L, I) 
    
class VGP_slow(VGP):
    def build_likelihood(self):
        K = self.kern.K(self.X)
        K_alpha = tf.matmul(K, self.q_alpha)
        f_mean = K_alpha + self.mean_function(self.X)
        K_inv = inv(K) # NN
        f_vars = [inv(K_inv + tf.diag(self.q_lambda[:, d]**2)) 
                                    for d in range(self.num_latent)] # D[NN]
        f_vars_diag = [tf.diag_part(f_var) for f_var in f_vars] # D[N]
        f_var = tf.transpose(tf.stack(f_vars_diag)) # ND
        
        q_sqrt = tf.transpose(tf.stack([tf.cholesky(f) for f in f_vars]), (1, 2, 0)) # NND 
        KL = GPflow.kullback_leiblers.gauss_kl(K_alpha, q_sqrt, K)
        v_exp = self.likelihood.variational_expectations(f_mean, f_var, self.Y)
        return tf.reduce_sum(v_exp) - KL 
```
This seems to be unusably unstable for more than about 50 points, but does appear to match at least for small N. e.g.

``` python
Ns = np.arange(1, 50)
for N in Ns:
    X = np.linspace(-10, 10, N)[:, None]
    Y = np.sin(X) + 0.9 * np.cos(X*1.6) + np.random.randn(*X.shape) * 0.8
    m1 = VGP(X, Y, RBF(1), Gaussian())
    m2 = VGP_slow(X, Y, RBF(1), Gaussian())

    q_alpha = np.random.randn(N, 1)
    q_lambda = np.random.randn(N, 1)**2

    for m in [m1, m2]:
        m.q_alpha = q_alpha
        m.q_lambda = q_lambda 
    L1 = m1.compute_log_likelihood()
    L2 = m2.compute_log_likelihood()
    print 'original: {:.4f}, slow: {:.4f}. Diff: {:.8f}'.format(L1, L2,  L1 - L2)
    
```

**Comment by alexggmatthews** [comment]
Fixed in #430
